clang/labs/lab1/balyasov_ivan/rename_identifiers.cpp

```diff
@@ -0,0 +1,193 @@
+      rewriter.ReplaceText(var->getLocation(), oldName.size(), newName);
+    }
+    if (type == Types::Class &&
+        var->getType().getAsString() == oldName + " *") {
+      rewriter.ReplaceText(var->getTypeSourceInfo()->getTypeLoc().getBeginLoc(),
+                           oldName.size(), newName);
+    }
+    if (type == Types::Class && var->getType().getAsString() == oldName) {
+      rewriter.ReplaceText(
+          var->getTypeSourceInfo()->getTypeLoc().getSourceRange(), newName);
+    }
+    return true;
+  }
+
+  bool VisitDeclRefExpr(clang::DeclRefExpr *expr) {
+    clang::VarDecl *var = clang::dyn_cast<clang::VarDecl>(expr->getDecl());
+    if (type == Types::Var && var && var->getName() == oldName) {
+      rewriter.ReplaceText(expr->getSourceRange(), newName);
+    }
+    return true;
+  }
+
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (type == Types::Class && record->getName() == oldName) {
+      rewriter.ReplaceText(record->getLocation(), newName);
+      const auto *destructor = record->getDestructor();
+      if (destructor) {
+        rewriter.ReplaceText(destructor->getLocation(), oldName.size() + 1,
+                             '~' + newName);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXConstructorDecl(clang::CXXConstructorDecl *constructor) {
+    if (type == Types::Class) {
+      if (constructor->getNameAsString() == oldName) {
+        rewriter.ReplaceText(constructor->getLocation(), oldName.size(),
+                             newName);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXNewExpr(clang::CXXNewExpr *newExpr) {
+    if (type == Types::Class) {
+      if (newExpr->getConstructExpr()->getType().getAsString() == oldName) {
+        rewriter.ReplaceText(newExpr->getExprLoc(), oldName.size() + 4,
+                             "new " + newName);
+      }
+    }
+    return true;
+  }
+
+  bool save_changes() { return rewriter.overwriteChangedFiles(); }
+};
+
+class RenameConsumer : public clang::ASTConsumer {
+private:
+  RenameVisitor Visitor;
+
+public:
+  explicit RenameConsumer(clang::CompilerInstance &CI, Types type,
+                          clang::StringRef oldName, clang::StringRef newName)
+      : Visitor(clang::Rewriter(CI.getSourceManager(), CI.getLangOpts()), type,
+                oldName, newName) {}
+
+  void HandleTranslationUnit(clang::ASTContext &context) override {
+    Visitor.TraverseDecl(context.getTranslationUnitDecl());
+    Visitor.save_changes();
+  }
+};
+
+class RenamePlugin : public clang::PluginASTAction {
+private:
+  Types type;
+  std::string oldName;
+  std::string newName;
+
+protected:
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &args) override {
+    std::vector<std::pair<std::string, std::string>> params = {
+        {"type=", ""}, {"cur-name=", ""}, {"new-name=", ""}};
+
+    if (!args.empty() && args[0] == "help") {
+      PrintHelp(llvm::errs());
+      return true;
```

Comment:
Why it returns true?

Reply:
True is returned because the arguments were processed without errors

Reply:
Yes, but will a plugin be executed after `help` in that case? Is it an expected behavior?

Reply:
Yes, it will be launched. I think this is unexpected behavior.

Reply:
Please fix it.

Reply:
I'm sorry, I was wrong. This is the expected behavior. Because we have the "help" flag processing.

Reply:
Yes, you are right, `help` flag is valid option, but when we execute an application and ask it to print a help, we don't expect that it starts to do something else, do we?

Reply:
Just a remainder that there is still a comment to fix.

Reply:
Okay, I think I understand.

Reply:
Thank you